### PR TITLE
Added Spectator, Death Roles | Added Rainbow Bed

### PR DESCRIPTION
**Pull request by xSavior-of-God**
Added Spectator and Death role to avoid giving the reward to those who die or to those who join as a spectator;
Also added the rainbow color bed for the "Leave" element ... works only if "material-name" is set to "RAINBOW_BED" (No CaSe SeNsItIvE!)... in short it gives a random color bed to each player

### CODE REVIEW

File: src/main/java/plugily/projects/murdermystery/arena/Arena.java [hunk 1]
```diff
@@ -80,7 +80,8 @@
   private final List<Stand> stands = new ArrayList<>();
   private final List<SpecialBlock> specialBlocks = new ArrayList<>();
   private final List<Player> allMurderer = new ArrayList<>(), allDetectives = new ArrayList<>();
-
+  private final List<Player> spectators = new ArrayList<>();
+  private final List<Player> deaths = new ArrayList<>();
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/arena/Arena.java:84]:
```suggestion
  private final List<Player> allMurderer = new ArrayList<>(), allDetectives = new ArrayList<>(), spectators = new ArrayList<>(), deaths = new ArrayList<>();

```
Keep cleaned-up the code

File: src/main/java/plugily/projects/murdermystery/arena/ArenaManager.java [hunk 2]
```diff
@@ -378,33 +381,42 @@ public static void stopGame(boolean quickStop, Arena arena) {
     List<String> summaryMessages = LanguageManager.getLanguageList("In-Game.Messages.Game-End-Messages.Summary-Message");
     arena.getScoreboardManager().stopAllScoreboards();
 
+    
+    
     boolean murderWon = arena.getPlayersLeft().size() == arena.aliveMurderer();
-
+
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/arena/ArenaManager.java:387]: Remove these whitespaces

File: src/main/java/plugily/projects/murdermystery/arena/ArenaManager.java [hunk 3]
```diff
@@ -321,6 +322,8 @@ public static void leaveAttempt(Player player, Arena arena) {
     }
     player.setGlowing(false);
     user.setSpectator(false);
+  	if(arena.isDeathPlayer(player)) arena.removeDeathPlayer(player);
+    if(arena.isSpectatorPlayer(player)) arena.removeSpectatorPlayer(player);
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/arena/ArenaManager.java:326]:
```suggestion
if (arena.isDeathPlayer(player)) {
  arena.removeDeathPlayer(player);
}

if (arena.isSpectatorPlayer(player)) {
  arena.removeSpectatorPlayer(player);
}

```
Keep the code readable

File: src/main/java/plugily/projects/murdermystery/arena/role/Role.java [hunk 4]
```diff
@@ -51,7 +52,15 @@
...
    * Murderer role, must kill everyone to win
    */
-  MURDERER;
+  MURDERER,
+  /**
+   * Spectator role, just look :D
+   */
+  SPECTATOR,
+  /**
+   * Death role, when everyone die
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/arena/role/Role.java:61]: die -> `died`

File: src/main/java/plugily/projects/murdermystery/handlers/items/SpecialItem.java [hunk 5]
```diff
@@ -67,20 +71,42 @@ public void load(String displayName, String[] lore, Material material, int slot)
...
-      .collect(Collectors.toList());
-    meta.setLore(colorizedLore);
-    stack.setItemMeta(meta);
-
-    SpecialItem item = new SpecialItem(name);
-    item.itemStack = stack;
-    item.slot = config.getInt(name + ".slot");
-    SpecialItemManager.addItem(name, item);
+    
+    if(config.getString(name + ".material-name").equalsIgnoreCase("RAINBOW_BED")) {
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/handlers/items/SpecialItem.java:75]: Add default key to `getString(path, default)`

File: src/main/java/plugily/projects/murdermystery/handlers/items/SpecialItemManager.java [hunk 6]
```diff
@@ -29,20 +31,22 @@
...
+  	Random num = new Random();
+  	return specialitem.get(num.nextInt(specialitem.size()));
   }
 
   public static String getRelatedSpecialItem(ItemStack itemStack) {
     for (String key : specialItems.keySet()) {
-      SpecialItem entityItem = specialItems.get(key);
-      if (entityItem.getItemStack().getItemMeta().getDisplayName().equalsIgnoreCase(itemStack.getItemMeta().getDisplayName())) {
+      List<SpecialItem> entityItem = specialItems.get(key);
+      if (entityItem.get(0).getItemStack().getItemMeta().getDisplayName().equalsIgnoreCase(itemStack.getItemMeta().getDisplayName())) {
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/handlers/items/SpecialItemManager.java:49]: `ArrayIndexOutOfBoundsException` can happen here, do a `!entityItem.isEmpty()` before `get`

File: src/main/java/plugily/projects/murdermystery/arena/Arena.java [hunk 7]
```diff
@@ -1040,5 +1043,29 @@ public void setBowHologram(ArmorStandHologram bowHologram) {
   public ArmorStandHologram getBowHologram() {
     return bowHologram;
   }
+  
+  public void addDeathPlayer(Player player) {
+  	this.deaths.add(player);
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/arena/Arena.java:1048]: Remove all keyword starts with `this.`

File: src/main/java/plugily/projects/murdermystery/arena/ArenaManager.java [hunk 8]
```diff
@@ -378,33 +381,42 @@ public static void stopGame(boolean quickStop, Arena arena) {
...
+						if (murderWon) {
+						  if (Role.isRole(Role.MURDERER, player)) {
+							  user.addStat(StatsStorage.StatisticType.WINS, 1);
+							  plugin.getRewardsHandler().performReward(player, Reward.RewardType.WIN);
+					  	} else {
+					  		user.addStat(StatsStorage.StatisticType.LOSES, 1);
+						    plugin.getRewardsHandler().performReward(player, Reward.RewardType.LOSE);
+						  }
+						} else {
+							if (!Role.isRole(Role.MURDERER, player)) {
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/arena/ArenaManager.java:406]: Chain else with if
```suggestion
						} else if (!Role.isRole(Role.MURDERER, player)) {
```

File: src/main/java/plugily/projects/murdermystery/arena/ArenaManager.java [hunk 9]
```diff
@@ -378,33 +381,42 @@ public static void stopGame(boolean quickStop, Arena arena) {
...
-            plugin.getRewardsHandler().performReward(player, Reward.RewardType.WIN);
-          } else {
-            user.addStat(StatsStorage.StatisticType.LOSES, 1);
-            plugin.getRewardsHandler().performReward(player, Reward.RewardType.LOSE);
-          }
-        }
-      }
+	  	User user = plugin.getUserManager().getUser(player);
+			if(!quickStop) {
+			  if (Role.isAnyRole(player)) {
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/arena/ArenaManager.java:391]:
```suggestion
			if (!quickStop && Role.isAnyRole(player)) {
```

**Comment by xSavior-of-God** [comment]
fail, sorry...

**Comment by montlikadani** [comment]
Don't worry, if you got time push the changes.

**Comment by xSavior-of-God** [comment]
I will make all the necessary changes... if there is nothing else, I will proceed 😁

File: src/main/java/plugily/projects/murdermystery/handlers/items/SpecialItem.java [hunk 10]
```diff
@@ -44,6 +45,8 @@
   private int slot;
   private final String name;
 
+  private static List<XMaterial> materialbed = new ArrayList<>(Arrays.asList(XMaterial.BLACK_BED, XMaterial.BLUE_BED, XMaterial.BROWN_BED, XMaterial.CYAN_BED, XMaterial.GRAY_BED, XMaterial.LIGHT_BLUE_BED, XMaterial.GREEN_BED, XMaterial.LIGHT_GRAY_BED, XMaterial.LIME_BED,XMaterial.MAGENTA_BED, XMaterial.ORANGE_BED, XMaterial.PINK_BED, XMaterial.PURPLE_BED, XMaterial.RED_BED, XMaterial.WHITE_BED, XMaterial.YELLOW_BED));
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/handlers/items/SpecialItem.java:48]:
```suggestion
  private final Set<XMaterial> beds = EnumSet.of(XMaterial.BLACK_BED, XMaterial.BLUE_BED, XMaterial.BROWN_BED,
   XMaterial.CYAN_BED, XMaterial.GRAY_BED, XMaterial.LIGHT_BLUE_BED, XMaterial.GREEN_BED, 
   XMaterial.LIGHT_GRAY_BED, XMaterial.LIME_BED,XMaterial.MAGENTA_BED, XMaterial.ORANGE_BED, 
   XMaterial.PINK_BED, XMaterial.PURPLE_BED, XMaterial.RED_BED, XMaterial.WHITE_BED, XMaterial.YELLOW_BED);
```

**Comment by xSavior-of-God** [comment]
done?

**Comment by montlikadani** [comment]
Seems other things are good, so you can push the changes.

**Comment by xSavior-of-God** [comment]
Done! 😁

File: src/main/java/plugily/projects/murdermystery/handlers/items/SpecialItem.java [hunk 11]
```diff
@@ -45,7 +45,14 @@
   private int slot;
   private final String name;
 
-  private static List<XMaterial> materialbed = new ArrayList<>(Arrays.asList(XMaterial.BLACK_BED, XMaterial.BLUE_BED, XMaterial.BROWN_BED, XMaterial.CYAN_BED, XMaterial.GRAY_BED, XMaterial.LIGHT_BLUE_BED, XMaterial.GREEN_BED, XMaterial.LIGHT_GRAY_BED, XMaterial.LIME_BED,XMaterial.MAGENTA_BED, XMaterial.ORANGE_BED, XMaterial.PINK_BED, XMaterial.PURPLE_BED, XMaterial.RED_BED, XMaterial.WHITE_BED, XMaterial.YELLOW_BED));
+  private static List<XMaterial> materialbed = new ArrayList<>(Arrays.asList(
```
Comment (montlikadani) [inline src/main/java/plugily/projects/murdermystery/handlers/items/SpecialItem.java:48]: Should be `Set<XMaterial>` and `EnumSet.of(XMaterial.BLACK_BED)`

**Review by montlikadani — APPROVED** [review]

**Comment by montlikadani** [comment]
Next time please do commits in `development` branch

**Pull request merged**